security/pkg/caclient/clientrunner.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package caclient
+
+import (
+	"fmt"
+	"time"
+
+	"istio.io/istio/pkg/log"
+	pkiutil "istio.io/istio/security/pkg/pki/util"
+	"istio.io/istio/security/pkg/util"
+)
+
+// ClientRunner interacts with the upstream CA to maintain the KeyCertBundle valid.
+type ClientRunner interface {
+	// Run the ClientRunner loop (blocking call)
+	Run(stopCh <-chan struct{}, errCh chan<- error)
+}
+
+type clientRunner struct {
+	keycert  pkiutil.KeyCertBundle
+	certUtil util.CertUtil
+	client   CAClient
+}
+
+// NewClientRunner creates a new clientReunner instance.
+func NewClientRunner(keycert pkiutil.KeyCertBundle, certUtil util.CertUtil, client CAClient) ClientRunner {
+	return &clientRunner{
+		keycert:  keycert,
+		certUtil: certUtil,
+		client:   client,
+	}
+}
+
+// Run periodically rotates the key/cert of the Istio CA by interacting with the upstream Istio CA.
+// It is a blocking function that should run as a go routine.
+func (c *clientRunner) Run(stopCh <-chan struct{}, errCh chan<- error) {
+	for {
```

Comment:
Will a bool channel be enough?

Reply:
nvm, I select the stopCh but seems like github's code review is does not support that...